base/applications/rapps/settings.cpp

```diff
@@ -0,0 +1,243 @@
+/*
+ * PROJECT:     ReactOS Applications Manager
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Functions to load / save settings from reg.
+ * COPYRIGHT:   Copyright 2020 He Yang           (1160386205@qq.com)
+ */
+
+#include "rapps.h"
+#include "settings.h"
+
+
+class SettingsField
+{
+public:
+    virtual ~SettingsField() { ; }
+    virtual BOOL Save(CRegKey &key) = 0;
+    virtual BOOL Load(CRegKey &key) = 0;
+};
+
+class SettingsFieldBool : public SettingsField
+{
+public:
+    SettingsFieldBool(BOOL *pValue, LPCWSTR szRegName)
+        : m_pValueStore(pValue), m_RegName(szRegName)
+    {
+    }
+
+    virtual BOOL Save(CRegKey &key) override
+    {
+        return key.SetDWORDValue(m_RegName, (DWORD)(*m_pValueStore)) == ERROR_SUCCESS;
+    }
+    virtual BOOL Load(CRegKey &key) override
+    {
+        DWORD dwField;
+        LONG lResult = key.QueryDWORDValue(m_RegName, dwField);
+        if (lResult != ERROR_SUCCESS)
+        {
+            return FALSE;
+        }
+        *m_pValueStore = (BOOL)dwField;
+        return TRUE;
+    }
+
+private:
+    BOOL *m_pValueStore;     // where to read/store the value
+    LPCWSTR m_RegName;       // key name in registery
+};
+
+class SettingsFieldInt : public SettingsField
+{
+public:
+    SettingsFieldInt(INT *pValue, LPCWSTR szRegName)
+        : m_pValueStore(pValue), m_RegName(szRegName)
+    {
+    }
+
+    virtual BOOL Save(CRegKey &key) override
+    {
+        return key.SetDWORDValue(m_RegName, (DWORD)(*m_pValueStore)) == ERROR_SUCCESS;
+    }
+    virtual BOOL Load(CRegKey &key) override
+    {
+        DWORD dwField;
+        LONG lResult = key.QueryDWORDValue(m_RegName, dwField);
+        if (lResult != ERROR_SUCCESS)
+        {
+            return FALSE;
+        }
+        *m_pValueStore = (INT)dwField;
+        return TRUE;
+    }
+
+private:
+    INT *m_pValueStore;      // where to read/store the value
+    LPCWSTR m_RegName;       // key name in registery
+};
+
+class SettingsFieldString : public SettingsField
+{
+public:
+    SettingsFieldString(WCHAR *pString, ULONG cchLen, LPCWSTR szRegName)
+        : m_pStringStore(pString), m_StringLen(cchLen), m_RegName(szRegName)
+    {
+    }
+
+    virtual BOOL Save(CRegKey &key) override
+    {
+        return key.SetStringValue(m_RegName, m_pStringStore) == ERROR_SUCCESS;
+    }
+    virtual BOOL Load(CRegKey &key) override
+    {
+        ULONG nChar = m_StringLen - 1; // make sure the terminating L'\0'
+        LONG lResult = key.QueryStringValue(m_RegName, m_pStringStore, &nChar);
```

Comment:
This one might overwrite the default value, but still return something else than ERROR_SUCCESS.